### ISSUE TITLE
Add a way to get the name of the implementation

### ISSUE THREAD

**Issue by Bammerbom** [SpongePowered/SpongeAPI#644]
Can someone add a way to get the name of the implementation?
There is currently a way to get the version, but not to get the name.

I need it for my own (private) version of mcstats

**Comment by MikaalSky** [comment]
What do you mean by implementation?

(A little bit newb at java :3)

**Comment by Bammerbom** [comment]
(I am not very good at explaining but I will try my best)

This is the Sponge API.
The api indicates what methods a certain class has. Like player#getGamemode(). The api contains only abstract classes.

The implementation indicates what the methods do.
Sponge has different implementations, such as:
- https://github.com/SpongePowered/Sponge
- https://github.com/SpongePowered/SpongeVanilla

**Comment by MikaalSky** [comment]
Oh so to detect if someone was running SpongeVanilla or just Sponge?

**Comment by MikaalSky** [comment]
That makes sense. I thought it might be that but I wasn't sure.

**Comment by Bammerbom** [comment]
Yes.

**Comment by MikaalSky** [comment]
This would also be useful for plugins that use things that arent implemented in a certain implementation.

**Comment by Azatik** [comment]
Something like:
```
server.getImplements().getName;
```
Return Sponge or SpongeVanilla.
???

**Comment by Bammerbom** [comment]
I was thinking about server.getImplementationName().

You get the version by server.getImplementationVersion() so that would be the most logical option.

**Comment by ST-DDT** [comment]
I agree with @Bammerbom  on this point.

[Game.getImplementationVersion()](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/Game.java#L121)

**Comment by ST-DDT** [comment]
In [SpongeCommon](https://github.com/SpongePowered/SpongeCommon/blob/master/src/main/java/org/spongepowered/common/SpongeGame.java#L51) we use [`Package.getImplementationVersion()`](http://docs.oracle.com/javase/7/docs/api/java/lang/Package.html#getImplementationVersion()).
Any objections for using [`Package.getImplementatioTitle()`](http://docs.oracle.com/javase/7/docs/api/java/lang/Package.html#getImplementationTitle())?
If not i could create a PR for both parts API and Impl.

**Comment by Bammerbom** [comment]
If that are Java standards, I agree with Package.getImplementationTitle()

**Comment by Mumfrey** [comment]
Personally I feel like if there is any method to return what is effectively platform metadata, ostensibly for debugging/reporting purposes (people making platform-specific behaviours should **not** receive support from the platform) then it should be in the form of some kind of `Platform` struct which contains information about the platform which can be easily extended without polluting another interface (like `Game`) with a lot of pointless methods. Since `Platform` in SpongeAPI is already used in a manner analogous to `Side` we could use `PlatformInfo`.

The intention would be that the full content of the `PlatformInfo` struct could then be included in crash reports and other troubleshooting output (such as log headers) to provide a consistent set of platform metadata. This also satisfies the above "statistics" goal, in that it would provide a known set of metadata for statistics systems to report.

I would suggest something like adding `Game.getPlatformInformation()` return a `PlatformInfo` with something like:

```java
public interface PlatformInfo {

    // These three methods moved from Game
    String getApiVersion();
    String getImplementationVersion();
    MinecraftVersion getMinecraftVersion();

    // New accessors
    String getImplementationName();
    OperatingSystem getOperatingSystem();

    // Other appropriate accessors
   ...

    // Summary, return all information as a KV map
    Map<String, ?> all();
```

This would be ideal for the ongoing fleshing out of error messages [as mentioned here](https://github.com/SpongePowered/Sponge/issues/257) since aggregating all the relevant data in a single place means that crash reporters only need to iterate over the map returned by `all()` and add all of the keys to their output.

**Comment by gabizou** [comment]
Pinging @SpongePowered/developers for agreement on @Mumfrey's suggestion above.

**Comment by Minecrell** [comment]
I think renaming the current `Platform` to `PlatformType` and then using `Platform` would make more sense and would be consistent with other parts of the API.

**Comment by DDoS** [comment]
Sounds good to me.

**Comment by Aaron1011** [comment]
This sounds like a good idea.

**Comment by MattBDev** [comment]
I agree with Mumfrey's suggestion.

**Comment by Minecrell** [comment]
If no one else is working on this already, I will add Mumfrey's suggestion with a few other ideas from myself and make a PR in the next days.

**Comment by gratimax** [comment]
Fixed by 42a5d59a
